### PR TITLE
Integrate the JPA map store

### LINKED ISSUE #13096: Integrate the JPA map store
### Description

See #12615.

This issue is about the initial changes to integrate the distribution with the JPA map store.

For now, we'll be leveraging the current database options (those within the database category and prefixed with `db`) to configure the JPA map store as we are going to initially support a single store for all areas.

It should also be possible to configure the server to run both `chm` and `jpa` stores at runtime. This is mainly needed to workaround some missing JPA-related providers like the one that persists `user sessions` so that we can use the `chm` store instead. It should also provide the baseline to support multiple stores simultaneously in the future.

### Discussion

_No response_

### Motivation

See #12615.

### Details

_No response_

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #13096

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java [hunk 1]
```diff
@@ -265,6 +265,7 @@ public void init(Config.Scope config) {
 
     @Override
     public void postInit(KeycloakSessionFactory factory) {
+        this.emf = createEntityManagerFactory();
     }
```
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java:270]: When testing this in Undertow and Quarkus, I found that this is being called twice (actually the logging in the undertow setup put me on this). 

Due to this, at least in the Undertow setup two EMFs will be set up, with two connection pools. I assume the same happens for Quarkus as well.

I remember this happens for all (?) `AmphibianProviderFactory`. This was apparently the reason why it was lazily initialized. 

Locally I rolled back the change and made it lazily-initializable. This then still worked for both Undertow and Quarkus.


```suggestion
    @Override
    public void postInit(KeycloakSessionFactory factory) {
    }

    // ... and then call lazyInit() again from create()
    private void lazyInit() {
        if (emf == null) {
            synchronized (this) {
                if (emf == null) {
                    this.emf = createEntityManagerFactory();
                }
            }
        }
    }

**Review by ahus1 — CHANGES_REQUESTED** [review]
Hi Pedro, thank you for this PR - this works really nice!

I was able to run this locally with the following command: 

```
bin/kc.sh start-dev --storage=jpa --db-password=pass --db-username=keycloak
```

When debugging/testing it, I found that the EMF is initialized twice, see the comment below. This is the only thing I'd like to have changed before merging it.

The following thing might be outside of the scope for this PR if there is no simple fix for it:

I also tried to run it with the maven plugin like this: 

```
mvn -f server/pom.xml compile quarkus:dev -Dquarkus.args="start-dev --storage=jpa --db-password=pass --db-username=keycloak"
```

unfortunately this failed with error messages about build time options: 

```
2022-07-14 09:50:36,518 WARN  [org.keycloak.common.Profile] (Quarkus Main Thread) Experimental feature enabled: map_storage
2022-07-14 09:50:36,570 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (Quarkus Main Thread) Hostname settings: FrontEnd: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin: <request>, Port: -1, Proxied: false
2022-07-14 09:50:37,007 INFO  [org.keycloak.common.crypto.CryptoIntegration] (Quarkus Main Thread) Detected security provider: org.keycloak.crypto.def.DefaultCryptoProvider
2022-07-14 09:50:37,352 WARN  [io.quarkus.runtime.configuration.ConfigRecorder] (Quarkus Main Thread) Build time property cannot be changed at runtime:
 - quarkus.datasource.jdbc.driver is set to 'org.postgresql.Driver' but it is build time fixed to 'org.h2.jdbcx.JdbcDataSource'. Did you change the property quarkus.datasource.jdbc.driver after building the application?
 - quarkus.datasource.jdbc.transactions is set to 'disabled' but it is build time fixed to 'xa'. Did you change the property quarkus.datasource.jdbc.transactions after building the application?
2022-07-14 09:50:37,366 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': No suitable driver found for jdbc:postgresql://localhost:5432/keycloak
2022-07-14 09:50:37,367 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: No suitable driver found for jdbc:postgresql://localhost:5432/keycloak
        at org.h2.jdbcx.JdbcDataSource.getJdbcConnection(JdbcDataSource.java:191)
```

I was expecting that the quarkus maven plugin would work the same as "start-dev", apparently it doesn't?

**Review by DGuhr — CHANGES_REQUESTED** [review]
Thanks @pedroigor for the PR, working nicely in general here. Also thanks @ahus1 for the review / catch with the different EMFs!

I Also tried it out and got the same error when running quarkus:dev than alex did - not 100% sure it is related to these changes here, though, it seems the build is not executed at all (even when you just use `...start-dev --db=postgres` or `start-dev --cache=ispn` it doesn't happen).

Here are some additional notes from my smoke test:
1) Just as a note: It seems storage jpa support is fixed to postgres for now, right? Should be documented later on.

2) Question: Why are changesets applied after startup? See the following log output: 
```
2022-07-14 11:59:11,168 INFO  [io.quarkus] (main) Keycloak 999-SNAPSHOT on JVM (powered by Quarkus 2.7.6.Final) started in 5.774s. Listening on: http://0.0.0.0:8080
2022-07-14 11:59:11,169 INFO  [io.quarkus] (main) Profile dev activated. 
2022-07-14 11:59:11,169 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql, keycloak, logging-gelf, narayana-jta, reactive-routes, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-health, smallrye-metrics, vault, vertx]
2022-07-14 11:59:11,172 WARN  [org.keycloak.quarkus.runtime.KeycloakMain] (main) Running the server in development mode. DO NOT use this configuration in production.
2022-07-14 12:03:11,783 INFO  [org.keycloak.services] (executor-thread-2) KC-SERVICES0077: Created initial admin user with username admin
2022-07-14 12:05:35,526 INFO  [org.keycloak.models.map.storage.jpa.liquibase.updater.MapJpaLiquibaseUpdaterProvider] (executor-thread-6) Updating database. Using changelog META-INF/jpa-auth-sessions-changelog.xml
Running Changeset: META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml::auth-sessions-1::keycloak
2022-07-14 12:05:37,685 INFO  [org.keycloak.models.map.storage.jpa.liquibase.updater.MapJpaLiquibaseUpdaterProvider] (executor-thread-5) Updating database. Using changelog META-INF/jpa-single-use-objects-changelog.xml
Running Changeset: META-INF/single-use-objects/jpa-single-use-objects-changelog-1.xml::single-use-object-1::keycloak
2022-07-14 12:10:55,143 INFO  [org.keycloak.models.map.storage.jpa.liquibase.updater.MapJpaLiquibaseUpdaterProvider] (executor-thread-1) Updating database. Using changelog META-INF/jpa-authz-changelog.xml
Running Changeset: META-INF/authz/resource-server/jpa-authz-resource-server-changelog-1.xml::authz-resource-server-1::keycloak
Running Changeset: META-INF/authz/scope/jpa-authz-scope-changelog-1.xml::authz-scope-1::keycloak
Running Changeset: META-INF/authz/resource/jpa-authz-resource-changelog-1.xml::authz-resource-1::keycloak
Running Changeset: META-INF/authz/resource/jpa-authz-resource-changelog-1.xml::authz-resource-2::keycloak
Running Changeset: META-INF/authz/policy/jpa-authz-policy-changelog-1.xml::authz-policy-1::keycloak
Running Changeset: META-INF/authz/policy/jpa-authz-policy-changelog-1.xml::authz-policy-2::keycloak
Running Changeset: META-INF/authz/permission/jpa-authz-permission-changelog-1.xml::authz-permission-1::keycloak
```
Is this intended?

Follow-up Question: This "Running Changeset" is not using a logger or anything and I can't find it in the code. Where does it originate from?

3)  Can we get rid of this warning when starting up ? Are we missing the resourcesVersionSeed?
```
2022-07-14 11:59:06,473 WARN  [org.keycloak.models.DeploymentStateProviderFactory] (main) It is recommended to set 'resourcesVersionSeed' property in the map provider config of deploymentState SPI
```

**Comment by ahus1** [comment]
> Why are changesets applied after startup? 

Those changesets are applied as needed for an area when it is touched for the first time, that is: when it is touching realms, the realms changeset runs. The same applies for all other areas. Eventually an area will be configured per realm; then those will be run when the realm is being touched for the first time.

> This "Running Changeset" is not using a logger or anything and I can't find it in the code. Where does it originate from?

I suppose this is from Liquibase, and Liquibase might have an SPI to discover the logging handler. If that fails, it might fallback to the console. I had assumed that this is "normal" with Quarkus, but maybe not?

**Review by ahus1 — DISMISSED** [review]
Thanks - this looks good to me! Thanks for preparing all this!

**Review by DGuhr — DISMISSED** [review]

**Comment by abstractj** [comment]
@pedroigor it needs a rebase, but please ping me afterward so I can merge.

**Review by ahus1 — APPROVED** [review]

**Comment by ahus1** [comment]
BTW: I found out why Liquibase logs to the console; will prepare another PR tomorrow.

**Pull request merged**